File: .claude/skills/claudex-routing/scripts/opencode_go_budget.py
#!/usr/bin/env python3
"""Evaluate the published OpenCode Go request budget for one usage window.

OpenCode Go publishes an *estimated* request allowance for each model and
window.  CodexBar exposes that provider's current window as ``usedPercent``;
it does not expose a request counter.  This module keeps the two facts
separate and reports an estimate instead of pretending that a local counter
is authoritative.
"""

from __future__ import annotations

import math
from typing import Any, Mapping

DEFAULT_PROVIDER = "opencode-go"
DEFAULT_MODEL = "opencode-go/deepseek-v4-flash"
DEFAULT_USAGE_PROVIDER = "opencodego"
DEFAULT_USAGE_WINDOW = "primary"
DEFAULT_WINDOW_MINUTES = 5 * 60
DEFAULT_ESTIMATED_REQUESTS = 31_650


def valid_request_budget(value: Any) -> bool:
    """Validate the config object without accepting implicit or unsafe values."""
    if not isinstance(value, Mapping):
        return False
    if set(value) != {"estimatedRequests", "windowMinutes", "usageWindow"}:
        return False
    requests = value.get("estimatedRequests")
    minutes = value.get("windowMinutes")
    window = value.get("usageWindow")
    return (
        isinstance(requests, int)
        and not isinstance(requests, bool)
        and requests > 0
        and isinstance(minutes, int)
        and not isinstance(minutes, bool)
        and minutes > 0
        and isinstance(window, str)
        and bool(window)
        and window.isascii()
        and all("!" <= char <= "~" for char in window)
    )


def normalized_request_budget(value: Any) -> dict[str, Any] | None:
    """Return a copy of a valid budget config, or ``None`` for no budget."""
    if not valid_request_budget(value):
        return None
    # Mapping values are copied so callers cannot mutate provider config via
    # the sanitized routing result.
    return {
        "estimatedRequests": int(value["estimatedRequests"]),
        "windowMinutes": int(value["windowMinutes"]),
        "usageWindow": str(value["usageWindow"]),
    }


def _status(
    available: bool,
    used_percent: float | None,
    reason: str,
    budget: Mapping[str, Any],
    **details: Any,
) -> dict[str, Any]:
    configured = normalized_request_budget(budget)
    if configured is None:
        raise ValueError("invalid OpenCode Go request budget")
    result: dict[str, Any] = {
        "available": available,
        "max_used_percent": used_percent,
        "remaining_percent": (
            None if used_percent is None else max(0.0, 100.0 - used_percent)
        ),
        "reason": reason,
        "request_budget": {
            "estimated_requests": configured["estimatedRequests"],
            "window_minutes": configured["windowMinutes"],
            "usage_window": configured["usageWindow"],
            "known": used_percent is not None,
            "used_percent": used_percent,
            **details,
        },
    }
    return result


def _valid_percent(value: Any) -> bool:
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and math.isfinite(float(value))
        and 0.0 <= float(value) <= 100.0
    )


def _entry(report: Any, provider: str) -> Mapping[str, Any] | None:
    if not isinstance(report, list):
        return None
    wanted = provider.casefold()
    for item in report:
        if (
            isinstance(item, Mapping)
            and str(item.get("provider", "")).casefold() == wanted
        ):
            return item
    return None


def evaluate(
    report: Any,
    usage_provider: str = DEFAULT_USAGE_PROVIDER,
    budget: Mapping[str, Any] | None = None,
) -> dict[str, Any] | None:
    """Evaluate a configured OpenCode Go budget against a CodexBar report.

    ``None`` means the provider has no request-budget configuration and the
    caller should use its normal quota handling.  A configured budget is
    considered unavailable when the provider/window/counter cannot be
    validated; routing must not silently fall back to another usage window.
    """
    normalized = normalized_request_budget(budget)
    if normalized is None:
        if budget is None:
            return None
        raise ValueError("invalid OpenCode Go request budget")

    entry = _entry(report, usage_provider)
    if entry is None:
        return _status(False, None, "missing", normalized)
    usage = entry.get("usage")
    window_name = normalized["usageWindow"]
    window = usage.get(window_name) if isinstance(usage, Mapping) else None
    if not isinstance(window, Mapping):
        return _status(False, None, "request-budget-window-missing", normalized)

    used_percent = window.get("usedPercent")
    reported_minutes = window.get("windowMinutes")
    if not _valid_percent(used_percent):
        return _status(False, None, "request-budget-usage-unknown", normalized)
    if (
        not isinstance(reported_minutes, int)
        or isinstance(reported_minutes, bool)
        or reported_minutes != normalized["windowMinutes"]
    ):
        return _status(
            False,
            None,
            "request-budget-window-mismatch",
            normalized,
            reported_window_minutes=reported_minutes,
        )

    percent = float(used_percent)
    total = float(normalized["estimatedRequests"])
    estimated_used = round(total * percent / 100.0, 3)
    estimated_remaining = round(max(0.0, total - estimated_used), 3)
    reset_at = window.get("resetsAt")
    if not isinstance(reset_at, str) or not reset_at:
        reset_at = None
    return _status(
        percent < 100.0,
        percent,
        "available" if percent < 100.0 else "request-budget-exhausted",
        normalized,
        reported_window_minutes=reported_minutes,
        estimated_used_requests=estimated_used,
        estimated_remaining_requests=estimated_remaining,
        resets_at=reset_at,
    )


def provider_budget(provider: Mapping[str, Any]) -> dict[str, Any] | None:
    """Read and validate the provider's optional request-budget policy."""
    value = provider.get("requestBudget")
    return normalized_request_budget(value) if value is not None else None


__all__ = [
    "DEFAULT_ESTIMATED_REQUESTS",
    "DEFAULT_MODEL",
    "DEFAULT_PROVIDER",
    "DEFAULT_USAGE_PROVIDER",
    "DEFAULT_USAGE_WINDOW",
    "DEFAULT_WINDOW_MINUTES",
    "evaluate",
    "normalized_request_budget",
    "provider_budget",
    "valid_request_budget",
]

File: tools/claudex-route-usage/src/opencode_go_budget.rs
//! Evaluate the published OpenCode Go request budget for one usage window.

use crate::util::{number_f64, python_round, valid_percentage};
use serde_json::{Map, Value};

pub fn evaluate(
    report: &Value,
    usage_provider: &str,
    budget: Option<&Value>,
) -> Result<Option<Value>, anyhow::Error> {
    let Some(budget_value) = budget else {
        return Ok(None);
    };
    let Some(normalized) = normalized_request_budget(budget_value) else {
        anyhow::bail!("invalid OpenCode Go request budget");
    };
    let Some(entry) = find_entry(report, usage_provider) else {
        return Ok(Some(status(
            false,
            None,
            "missing",
            &normalized,
            Map::new(),
        )));
    };
    let window_name = normalized["usageWindow"].as_str().unwrap_or_default();
    let window = entry
        .get("usage")
        .and_then(Value::as_object)
        .and_then(|usage| usage.get(window_name));
    let Some(window) = window.filter(|value| value.is_object()) else {
        return Ok(Some(status(
            false,
            None,
            "request-budget-window-missing",
            &normalized,
            Map::new(),
        )));
    };
    let used_percent = window.get("usedPercent");
    let reported_minutes = window.get("windowMinutes");
    if used_percent.is_none_or(|value| !valid_budget_percent(value)) {
        return Ok(Some(status(
            false,
            None,
            "request-budget-usage-unknown",
            &normalized,
            Map::new(),
        )));
    }
    let expected_minutes = normalized["windowMinutes"].as_i64();
    let reported_ok = reported_minutes
        .filter(|value| !value.is_boolean())
        .and_then(Value::as_i64)
        .zip(expected_minutes)
        .is_some_and(|(reported, expected)| reported == expected);
    if !reported_ok {
        let mut details = Map::new();
        details.insert(
            "reported_window_minutes".into(),
            reported_minutes.cloned().unwrap_or(Value::Null),
        );
        return Ok(Some(status(
            false,
            None,
            "request-budget-window-mismatch",
            &normalized,
            details,
        )));
    }
    let percent = number_f64(used_percent.unwrap()).unwrap_or_default();
    let total = normalized["estimatedRequests"].as_f64().unwrap_or_default();
    let estimated_used = python_round(total * percent / 100.0, 3);
    let estimated_remaining = python_round((total - estimated_used).max(0.0), 3);
    let reset_at = window
        .get("resetsAt")
        .and_then(Value::as_str)
        .filter(|text| !text.is_empty())
        .map_or(Value::Null, Value::from);
    let mut details = Map::new();
    details.insert(
        "reported_window_minutes".into(),
        reported_minutes.cloned().unwrap_or(Value::Null),
    );
    details.insert(
        "estimated_used_requests".into(),
        Value::from(estimated_used),
    );
    details.insert(
        "estimated_remaining_requests".into(),
        Value::from(estimated_remaining),
    );
    details.insert("resets_at".into(), reset_at);
    Ok(Some(status(
        percent < 100.0,
        Some(percent),
        if percent < 100.0 {
            "available"
        } else {
            "request-budget-exhausted"
        },
        &normalized,
        details,
    )))
}

fn normalized_request_budget(value: &Value) -> Option<Map<String, Value>> {
    if !crate::config::valid_request_budget(value) {
        return None;
    }
    let object = value.as_object()?;
    let mut out = Map::new();
    out.insert(
        "estimatedRequests".into(),
        Value::from(object["estimatedRequests"].as_i64().unwrap_or_default()),
    );
    out.insert(
        "windowMinutes".into(),
        Value::from(object["windowMinutes"].as_i64().unwrap_or_default()),
    );
    out.insert(
        "usageWindow".into(),
        Value::from(object["usageWindow"].as_str().unwrap_or_default()),
    );
    Some(out)
}

fn find_entry<'a>(report: &'a Value, provider: &str) -> Option<&'a Value> {
    let entries = report.as_array()?;
    entries.iter().find(|item| {
        item.get("provider")
            .and_then(Value::as_str)
            .is_some_and(|name| name.eq_ignore_ascii_case(provider))
    })
}

fn valid_budget_percent(value: &Value) -> bool {
    valid_percentage(value) && number_f64(value).is_some_and(|n| n <= 100.0)
}

fn status(
    available: bool,
    used_percent: Option<f64>,
    reason: &str,
    budget: &Map<String, Value>,
    details: Map<String, Value>,
) -> Value {
    let mut request_budget = Map::new();
    request_budget.insert(
        "estimated_requests".into(),
        budget["estimatedRequests"].clone(),
    );
    request_budget.insert("window_minutes".into(), budget["windowMinutes"].clone());
    request_budget.insert("usage_window".into(), budget["usageWindow"].clone());
    request_budget.insert("known".into(), Value::from(used_percent.is_some()));
    request_budget.insert(
        "used_percent".into(),
        used_percent.map_or(Value::Null, Value::from),
    );
    for (key, value) in details {
        request_budget.insert(key, value);
    }
    serde_json::json!({
        "available": available,
        "max_used_percent": used_percent,
        "remaining_percent": used_percent.map(|used| (100.0 - used).max(0.0)),
        "reason": reason,
        "request_budget": request_budget,
    })
}
